### PR TITLE
support insecure client for internal components

### LINKED ISSUE #4251: Support not setting `security.ssl-ca` when `spec.tidb.tlsclient.enabled: true`
## Feature Request

**Is your feature request related to a problem? Please describe:**

Currently, when the user set `spec.tidb.tlsclient.enabled: true` for TiDB, the following configurations will be set automatically [here](https://github.com/pingcap/tidb-operator/blob/v1.2.4/pkg/manager/member/tidb_member_manager.go#L388):
```
      [security]
        ssl-ca = "/var/lib/tidb-server-tls/ca.crt"
        ssl-cert = "/var/lib/tidb-server-tls/tls.crt"
        ssl-key = "/var/lib/tidb-server-tls/tls.key"
```

However, according to the doc [here](https://docs.pingcap.com/tidb/stable/enable-tls-between-clients-and-servers#configure-tidb-server-to-use-secure-connections), the `ssl-ca` is optional if the user does not need to enable the authentication between TiDB client and TiDB server.
So we have to support the above behavior.
**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by just1900**
### What problem does this PR solve?
Support insecure client for internal components
Fix #4251
### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [x] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
support insecure client for internal components 
```

### CODE REVIEW

File: pkg/manager/member/pd_member_manager.go [hunk 1]
```diff
@@ -829,7 +829,9 @@ func getPDConfigMap(tc *v1alpha1.TidbCluster) (*corev1.ConfigMap, error) {
 	}
 	// Versions below v4.0 do not support Dashboard
 	if tc.Spec.TiDB != nil && tc.Spec.TiDB.IsTLSClientEnabled() && !tc.SkipTLSWhenConnectTiDB() && clusterVersionGE4 {
-		config.Set("dashboard.tidb-cacert-path", path.Join(tidbClientCertPath, tlsSecretRootCAKey))
+		if !tc.Spec.TiDB.TLSClient.DisableInternalClientVerify {
+			config.Set("dashboard.tidb-cacert-path", path.Join(tidbClientCertPath, tlsSecretRootCAKey))
```
Comment (csuzhangxc) [inline pkg/manager/member/pd_member_manager.go:833]: only for Dashboard, not `internal components like Initializer, Dashboard, etc`

File: cmd/backup-manager/app/cmd/restore.go [hunk 2]
```diff
@@ -47,6 +47,7 @@ func NewRestoreCommand() *cobra.Command {
 	cmd.Flags().StringVar(&ro.TiKVVersion, "tikvVersion", util.DefaultVersion, "TiKV version")
 	cmd.Flags().BoolVar(&ro.TLSClient, "client-tls", false, "Whether client tls is enabled")
 	cmd.Flags().BoolVar(&ro.TLSCluster, "cluster-tls", false, "Whether cluster tls is enabled")
+	cmd.Flags().BoolVar(&ro.InsecureSkipVerify, "insecure", false, "Whether to skip tidb server's certificates validation")
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/cmd/restore.go:50]:
```suggestion
	cmd.Flags().BoolVar(&ro.InsecureSkipVerify, "skipClientCA", false, "Whether to skip tidb server's certificates validation")
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 3]
```diff
@@ -1286,10 +1286,15 @@ type TiDBTLSClient struct {
...
-	// DisableClientAuthn will skip client authentication from the TiDB server.
+	// DisableClientAuthn will skip client's certificate validation from the TiDB server.
 	// Optional: defaults to false
 	// +optional
 	DisableClientAuthn bool `json:"disableClientAuthn,omitempty"`
+
+	// DisableInternalClientVerify will skip TiDB server's certificate validation for internal components like Initializer, Dashboard, etc.
+	// Optional: defaults to false
+	// +optional
+	DisableInternalClientVerify bool `json:"disableInternalClientVerify,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1297]:
```suggestion
	SkipInternalClientCA bool `json:"skipInternalClientCA,omitempty"`
```

**Review by csuzhangxc — APPROVED** [review]

File: cmd/backup-manager/app/import/import.go [hunk 4]
```diff
@@ -107,7 +107,9 @@ func (ro *Options) loadTidbClusterData(ctx context.Context, restorePath string,
 	}
 
 	if ro.TLSClient {
-		args = append(args, fmt.Sprintf("--ca=%s", path.Join(util.TiDBClientTLSPath, corev1.ServiceAccountRootCAKey)))
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/import/import.go:110]: What about backup/restore/export?

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 5]
```diff
@@ -1339,192 +1339,226 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+				ginkgo.By("Inserting data into source db")
+				err = wait.PollImmediate(time.Second*5, time.Minute*5, insertIntoDataToSourceDB(fw, c, ns, tcName, passwd, true))
+				framework.ExpectNoError(err, "insert data into source db timeout")
+
+				ginkgo.By("Checking tidb-binlog works as expected")
+				err = wait.PollImmediate(time.Second*5, time.Minute*5, dataInClusterIsCorrect(fw, c, ns, targetTcName, "", false))
+				framework.ExpectNoError(err, "check data correct timeout")
+
+				ginkgo.By("Connecting to tidb server to verify the connection is TLS enabled")
+				err = wait.PollImmediate(time.Second*5, time.Minute*5, tidbIsTLSEnabled(fw, c, ns, tcName, passwd))
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1518]: It seems that `tidbIsTLSEnabled` will use the ca.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 6]
```diff
@@ -1339,192 +1339,226 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
-			framework.ExpectNoError(err, "failed to update TidbCluster: %q", tc.Name)
-			err = oa.WaitForTidbClusterReady(tc, 30*time.Minute, 5*time.Second)
-			framework.ExpectNoError(err, "wait for TidbCluster ready timeout: %q", tc.Name)
+				ginkgo.By("Installing tidb components certificates")
+				err = InstallTiDBComponentsCertificates(ns, tcName)
+				framework.ExpectNoError(err, "failed to install tidb components certificates")
 
-			ginkgo.By("Scaling in tidb cluster")
-			err = controller.GuaranteedUpdate(genericCli, tc, func() error {
+				ginkgo.By("Removing ca.crt in tidb-client-secret")
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1383]: Should also remove the ca from the secrets of initializer and dashboards when skipCA is set?

File: cmd/backup-manager/app/import/import.go [hunk 7]
```diff
@@ -107,7 +107,9 @@ func (ro *Options) loadTidbClusterData(ctx context.Context, restorePath string,
...
+		if !ro.InsecureSkipVerify {
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/import/import.go:110]: Does this also need to update?

File: pkg/backup/backup/backup_manager.go [hunk 8]
```diff
@@ -406,6 +406,10 @@ func (bm *backupManager) makeBackupJob(backup *v1alpha1.Backup) (*batchv1.Job, s
 
 	if backup.Spec.From != nil && tc.Spec.TiDB != nil && tc.Spec.TiDB.TLSClient != nil && tc.Spec.TiDB.TLSClient.Enabled && !tc.SkipTLSWhenConnectTiDB() {
 		args = append(args, "--client-tls=true")
+		if tc.Spec.TiDB.TLSClient.SkipInternalClientCA {
```
Comment (DanielZhangQD) [inline pkg/backup/backup/backup_manager.go:409]: Also need to update `makeExportJob`.

File: pkg/backup/restore/restore_manager.go [hunk 9]
```diff
@@ -379,6 +379,10 @@ func (rm *restoreManager) makeRestoreJob(restore *v1alpha1.Restore) (*batchv1.Jo
 
 	if restore.Spec.To != nil && tc.Spec.TiDB != nil && tc.Spec.TiDB.TLSClient != nil && tc.Spec.TiDB.TLSClient.Enabled && !tc.SkipTLSWhenConnectTiDB() {
 		args = append(args, "--client-tls=true")
+		if tc.Spec.TiDB.TLSClient.SkipInternalClientCA {
```
Comment (DanielZhangQD) [inline pkg/backup/restore/restore_manager.go:382]: Also need to update `makeImportJob`.

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**